spring-kafka/src/main/java/org/springframework/kafka/core/DefaultTransactionSuffixManager.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright 2016-2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.kafka.core;
+
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import org.springframework.lang.Nullable;
+import org.springframework.util.Assert;
+
+/**
+ * The {@link TransactionSuffixManager} implementation for managing transactional producer suffixes.
+ * If the {@link #maxCache} is greater than 0, the suffixes will be cached and reused.
+ * Otherwise, the suffixes will be generated on the fly.
+ *
+ * <p>
+ * Setting the {@link #setMaxCache(int)} enables caching to restrict the number of `transactional.id`.
+ * The cache is per `transactional.id` prefix. The cache size is limited by the {@link #maxCache}.
+ * Default is 0, which means no caching and restriction, so the `transactional.id` will be generated on the fly.
+ *
+ * @author Ilya Starchenko
+ *
+ * @since 3.1.1
+ */
+public class DefaultTransactionSuffixManager implements TransactionSuffixManager {
+
+	private final AtomicInteger transactionIdSuffix = new AtomicInteger();
+
+	private final Map<String, BlockingQueue<String>> suffixCache = new ConcurrentHashMap<>();
+
+	private int maxCache;
+
+	/**
+	 * Retrieve the suffix for the transactional producer from the cache or generate a new one 
+	 * if caching is disabled.
+	 * @param txIdPrefix the transaction id prefix.
+	 * @return the suffix.
+	 * @throws NoProducerAvailableException if caching is enabled and no suffixes are available.
+	 */
+	@Override
+	public String retrieveSuffix(String txIdPrefix) {
+		BlockingQueue<String> cache = getSuffixCache(txIdPrefix);
+		if (cache == null) {
```

Comment:
This may never be null since the value is computed in the `getSuffixCache` method.

Reply:
It could be null if `txIdPrefix` is not passed or caching is not enabled(maxCache <= 0).